usr/sbin/p11sak/p11sak.c

```diff
@@ -1908,106 +2053,277 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
                 printf(" %*s | ", CELL_SIZE, keytype);
             printf("%s\n", label);
         }
+        free(label);
+        free(keytype);
     }
 
-    ret = funcs->C_FindObjectsFinal(session);
-    if (ret != CKR_OK) {
-        printf("C_FindObjectsFinal failed (error code 0x%lX)\n", ret);
+    rc = funcs->C_FindObjectsFinal(session);
+    if (rc != CKR_OK) {
+        printf("C_FindObjectsFinal failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
         goto done;
     }
 
-    ret = CKR_OK;
+    rc = CKR_OK;
+
+done:
+
+    return rc;
+}
+
+static CK_RV getKeyboardInput(char *buff, char *label)
+{
+
+    CK_RV rc = CKR_OK;
+    int nread;
+    size_t buflen;
+
+    printf("Are you sure to delete %s? (y/n)\n", label);
+
+    nread = getline(&buff, &buflen, stdin);
```

Comment:
I don't think that this works as intended. 
getline() allocates  buff, if buff is NULL. If buff is not NULL, it checks if buflen is large enough to hold the input. If it is not large enough then it will reallocate the buffer.
You do allocate buff in the calling routine, but only in 1 byte length:
`if (!(buff = malloc(strlen("y"))))`
Thats wrong in 3 places:
1. strlen("y") returns 1, but you also need space for the terminating zero.
2. buflen is not initialized, so getline will get garbage as buffer length.
3. What if the user input is larger than one character? Then getline will reallocate buff, but you don't pass it back to the caller.